plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java

```diff
@@ -1705,4 +1705,58 @@ private void testWriteWithFileSize(String testTable, int scaleFactorInThousands,
                         .row(true)
                         .build());
     }
+
+    @Test
+    public void testAnalyzePartitionedTableWithCanonicalization()
+    {
+        String tableName = "test_analyze_table_canonicalization_" + randomNameSuffix();
+        assertUpdate(
+                """
+                CREATE TABLE %s (a_varchar varchar, month varchar)
+                WITH (
+                    partitioned_by = ARRAY['month'])
+                """.formatted(getFullyQualifiedTestTableName(tableName)));
+
+        assertUpdate("INSERT INTO " + getFullyQualifiedTestTableName(tableName) + " VALUES ('A', '01'), ('AA', '01'), ('B', '02'), ('M', '10')", 4);
+
+        String tableLocation = (String) computeActual("SELECT DISTINCT regexp_replace(\"$path\", '/[^/]*/[^/]*$', '') FROM " + getFullyQualifiedTestTableName(tableName)).getOnlyValue();
+
+        String externalTableName = "external_" + tableName;
+        assertUpdate(format(
+                """
+                CREATE TABLE %s(
+                  a_varchar varchar,
+                  month integer)
+                WITH (
+                   partitioned_by = ARRAY['month'],
+                   external_location='%s')
+                """.formatted(getFullyQualifiedTestTableName(externalTableName), tableLocation)));
+        assertUpdate(format("CALL system.sync_partition_metadata('%s', '%s', 'ADD', false)", HIVE_TEST_SCHEMA, externalTableName));
```

Comment:
we're relying on sync_partition_metadata behavior, which is implicit & subject to change
in fact, sync_partition_metadata could canonicalize the values inferred from the storage, why not

there should be some `onHive` way to enter into this non-canonical partitions state.
Or, we could use HMS client directly (product tests can do that)

Reply:
Sounds like https://github.com/trinodb/trino/pull/15938/ would be a good way to help out in such a situation.

Reply:
I rewrote the test to the form:

```
    @Test
    public void testAnalyzePartitionedTableWithCanonicalization()
    {
        String tableName = "test_analyze_table_canonicalization_" + randomNameSuffix();
        assertUpdate("CREATE TABLE %s (a_varchar varchar, month integer) WITH (partitioned_by = ARRAY['month'])".formatted(getFullyQualifiedTestTableName(tableName)));

        assertUpdate("INSERT INTO " + getFullyQualifiedTestTableName(tableName) + " VALUES ('A', 1), ('AA', 1), ('B', 2), ('M', 10)", 4);

        // Simulate the fact that the Hive table has non-canonical partition values
        hiveMinioDataLake.getHiveHadoop().runOnMetastore(
                """
                UPDATE PARTITIONS
                SET part_name='month=01'
                WHERE
                   TBL_ID IN (SELECT tbl_id FROM TBLS t INNER JOIN DBS db ON t.db_id = db.db_id WHERE db.name = '%s' and t.tbl_name = '%s') AND
                   PART_NAME='month=1'
                """.formatted(HIVE_TEST_SCHEMA, tableName));
        assertQuery("SELECT * FROM " + HIVE_TEST_SCHEMA + ".\"" + tableName + "$partitions\"", "VALUES 1, 2, 10");
        assertUpdate("ANALYZE " + getFullyQualifiedTestTableName(tableName), 4);
        assertQuery("SHOW STATS FOR " + getFullyQualifiedTestTableName(tableName),
                """
                        VALUES
                            ('a_varchar', 5.0, 2.0, 0.0, null, null, null),
                            ('month', null, 3.0, 0.0, null, 1, 10),
                            (null, null, null, null, 4.0, null, null)
                        """);

        assertUpdate("INSERT INTO " + getFullyQualifiedTestTableName(tableName) + " VALUES ('C', 3)", 1);
        hiveMinioDataLake.getHiveHadoop().runOnMetastore(
                """
                UPDATE PARTITIONS
                SET part_name='month=03'
                WHERE
                   TBL_ID IN (SELECT tbl_id FROM TBLS t INNER JOIN DBS db ON t.db_id = db.db_id WHERE db.name = '%s' and t.tbl_name = '%s') AND
                   PART_NAME='month=3'
                """.formatted(HIVE_TEST_SCHEMA, tableName));

        assertUpdate("ANALYZE " + getFullyQualifiedTestTableName(tableName) + " WITH (partitions = ARRAY[ARRAY['03']])", 1);
        assertQuery("SHOW STATS FOR " + getFullyQualifiedTestTableName(tableName),
                """
                        VALUES
                            ('a_varchar', 6.0, 2.0, 0.0, null, null, null),
                            ('month', null, 4.0, 0.0, null, 1, 10),
                            (null, null, null, null, 5.0, null, null)
                        """);
        // TODO (https://github.com/trinodb/trino/issues/15998) fix selective ANALYZE for table with non-canonical partition values
        assertQueryFails("ANALYZE " + getFullyQualifiedTestTableName(tableName) + " WITH (partitions = ARRAY[ARRAY['3']])", "Partition no longer exists: month=3");

        assertUpdate("DROP TABLE " + getFullyQualifiedTestTableName(tableName));
    }
```

but unfortunately we stumble on 

```
Caused by: io.trino.spi.TrinoException: Partition no longer exists: month=01
	at io.trino.plugin.hive.HiveSplitManager.lambda$getPartitionMetadata$3(HiveSplitManager.java:323)
	at com.google.common.collect.Iterators$6.transform(Iterators.java:829)
```

Probably related to #15998

Reply:
https://github.com/trinodb/trino/blob/0954991016d98487d1ff718ad7aa86ac5d48edb8/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java#L346-L347

we'd probably need here the `partitionTypes` to do something similar as in `io.trino.plugin.hive.HiveMetadata#canonicalizePartitionValues`

`"month=01"` -> would get us `"01"` to the raw partition value -> which would need to be parsed to `"1"`

Reply:
Changed the code to avoid depending  on `system.sync_partition_metadata` procedure call and instead add the partitions of the table manually through the metastore client.